hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -960,10 +961,23 @@ private TcpClientConnection onAuthenticated(TcpClientConnection connection,
                 // The first connection that opens a connection to the new cluster should set `clusterId`.
                 // This one will initiate `initializeClientOnCluster` if necessary.
                 clusterId = newClusterId;
-                if (clusterIdChanged) {
+                if (clusterIdChanged || establishedInitialClusterConnection) {
+                    // In split brain, the client might connect to the one half
+                    // of the cluster, and then later might reconnect to the
+                    // other half, after the half it was connected to is
+                    // completely dead. Since the cluster id is preserved in
+                    // split brain scenarios, it is impossible to distinguish
+                    // reconnection to the same cluster vs reconnection to the
+                    // other half of the split brain. However, in latter,
+                    // we might need to send some state to the other half of
+                    // the split brain (like Compact schemas or user code
+                    // deployment classes). That forces us to send client state
+                    // to the cluster after first cluster connection, regardless
+                    // the cluster id is changed or not.
                     clientState = ClientState.CONNECTED_TO_CLUSTER;
                     executor.execute(() -> initializeClientOnCluster(newClusterId));
```

Comment:
After the initial cluster connection, for all the member connections will we be executing this block of code?

Reply:
No, we will do that only if `connectionsEmpty`, meaning we are re-connecting to the cluster. See the if check above